README.md
# Pie Chart ![Cirrus CI - Base Branch Build Status](https://img.shields.io/cirrus/github/apgapg/caviare_flutter_pie_chart) [![GitHub stars](https://img.shields.io/github/stars/apgapg/caviare_flutter_pie_chart.svg?style=social)](https://github.com/apgapg/caviare_flutter_pie_chart) [![Twitter Follow](https://img.shields.io/twitter/url/https/@ayushpgupta.svg?style=social)](https://twitter.com/ayushpgupta) ![GitHub last commit](https://img.shields.io/github/last-commit/apgapg/caviare_flutter_pie_chart.svg) [![Website shields.io](https://img.shields.io/website-up-down-green-red/http/shields.io.svg)](https://apgapg.github.io/)[![Open Source Love](https://badges.frapsoft.com/os/v2/open-source.svg?v=103)](https://github.com/apgapg/caviare_flutter_pie_chart) 

This Flutter package provides a Pie Chart Widget with cool animation.

## 💻 Try LIVE Demo

Live Demo: [https://apgapg.github.io/caviare_flutter_pie_chart/](https://apgapg.github.io/caviare_flutter_pie_chart/)

<img src="https://raw.githubusercontent.com/apgapg/caviare_flutter_pie_chart/master/res/s10.png?raw=true"  width = "auto" alt="piechart">

## 💻 Installation
In the `dependencies:` section of your `pubspec.yaml`, add the following line:

[![Version](https://img.shields.io/pub/v/caviare_flutter_pie_chart.svg)](https://pub.dartlang.org/packages/caviare_flutter_pie_chart)

```yaml
dependencies:
  caviare_flutter_pie_chart: <latest version>
```

## ❔ Usage

### Import this class

```dart
import 'package:caviare_flutter_pie_chart/caviare_flutter_pie_chart.dart';
```

#### Usage is simple. Pie Chart is a widget and it just need a Map<String,double> as its data input.

<img src="https://raw.githubusercontent.com/apgapg/caviare_flutter_pie_chart/master/res/app.gif"  height = "400" alt="PieChart">

```dart
Map<String, double> dataMap = {
    "Flutter": 5,
    "React": 3,
    "Xamarin": 2,
    "Ionic": 2,
  };
```

### - Simple Implementation
```dart
PieChart(dataMap: dataMap) 
```

### - Full Implementation
```dart
PieChart(
      dataMap: dataMap,
      animationDuration: Duration(milliseconds: 800),
      chartLegendSpacing: 32,
      chartRadius: MediaQuery.of(context).size.width / 3.2,
      colorList: colorList,
      initialAngleInDegree: 0,
      chartType: ChartType.ring,
      ringStrokeWidth: 32,
      centerText: "HYBRID",
      legendOptions: LegendOptions(
        showLegendsInRow: false,
        legendPosition: LegendPosition.right,
        showLegends: true,
        legendShape: _BoxShape.circle,
        legendTextStyle: TextStyle(
          fontWeight: FontWeight.bold,
        ),
      ),
      chartValuesOptions: ChartValuesOptions(
        showChartValueBackground: true,
        showChartValues: true,
        showChartValuesInPercentage: false,
        showChartValuesOutside: false,
        decimalPlaces: 1,
      ),
      // gradientList: ---To add gradient colors---
      // emptyColorGradient: ---Empty Color gradient---
    )
```

### Control decimal places 'decimalPlaces'

```dart
chartValuesOptions: ChartValuesOptions(
  decimalPlaces: 1,
),
```

### Change legend position with 'legendPosition'

<img src="https://raw.githubusercontent.com/apgapg/caviare_flutter_pie_chart/master/res/s4.png"  height = "400" alt="PieChart"> <img src="https://raw.githubusercontent.com/apgapg/caviare_flutter_pie_chart/master/res/s5.png"  height = "400" alt="PieChart"> <img src="https://raw.githubusercontent.com/apgapg/caviare_flutter_pie_chart/master/res/s6.png"  height = "400" alt="PieChart"> <img src="https://raw.githubusercontent.com/apgapg/caviare_flutter_pie_chart/master/res/s7.png"  height = "400" alt="PieChart">

### Change Chart shape to ring

```dart
chartType: ChartType.ring,
```
<img src="https://raw.githubusercontent.com/apgapg/caviare_flutter_pie_chart/master/res/s9.png"  height = "400" alt="PieChart">

```dart
chartType: ChartType.ring,
showChartValuesOutside: true,
```

<img src="https://raw.githubusercontent.com/apgapg/caviare_flutter_pie_chart/master/res/s8.png"  height = "400" alt="PieChart">

### Gradient Colors

Pie Chart supports gradient as colors. Just pass `gradientList` instead of `colorList` to add gradient to chart colors.

```dart
final gradientList = <List<Color>>[
  [
    Color.fromRGBO(223, 250, 92, 1),
    Color.fromRGBO(129, 250, 112, 1),
  ],
  [
    Color.fromRGBO(129, 182, 205, 1),
    Color.fromRGBO(91, 253, 199, 1),
  ],
  [
    Color.fromRGBO(175, 63, 62, 1.0),
    Color.fromRGBO(254, 154, 92, 1),
  ]
];

// Pass gradient to PieChart
gradientList: gradientList,
emptyColorGradient: [
 Color(0xff6c5ce7),
 Colors.blue,
],
```

<img src="https://raw.githubusercontent.com/apgapg/caviare_flutter_pie_chart/master/res/s11.png"  height = "400" alt="PieChart">

## ⭐ My Flutter Packages
- [json_table](https://pub.dartlang.org/packages/json_table)  [![GitHub stars](https://img.shields.io/github/stars/apgapg/json_table.svg?style=social)](https://github.com/apgapg/json_table)  Create Flutter Json Table from json map directly.
- [avatar_glow](https://pub.dartlang.org/packages/avatar_glow)  [![GitHub stars](https://img.shields.io/github/stars/apgapg/avatar_glow.svg?style=social)](https://github.com/apgapg/avatar_glow)  Flutter Avatar Glow Widget with glowing animation.
- [search_widget](https://pub.dartlang.org/packages/search_widget)  [![GitHub stars](https://img.shields.io/github/stars/apgapg/search_widget.svg?style=social)](https://github.com/apgapg/search_widget)  Flutter Search Widget for selecting an option from list.
- [animating_location_pin](https://pub.dev/packages/animating_location_pin)  [![GitHub stars](https://img.shields.io/github/stars/apgapg/animating_location_pin.svg?style=social)](https://github.com/apgapg/animating_location_pin)  Flutter Animating Location Pin Widget providing Animating Location Pin Widget which can be used while fetching device location.
                                                                                                                                                                                                                             
## ⭐ My Flutter Apps
- [flutter_profile](https://github.com/apgapg/flutter_profile)  [![GitHub stars](https://img.shields.io/github/stars/apgapg/flutter_profile.svg?style=social)](https://github.com/apgapg/flutter_profile)  Showcase My Portfolio: Ayush P Gupta on Playstore.
- [flutter_sankalan](https://github.com/apgapg/flutter_sankalan)  [![GitHub stars](https://img.shields.io/github/stars/apgapg/flutter_sankalan.svg?style=social)](https://github.com/apgapg/flutter_sankalan)  Flutter App which allows reading/uploading short stories.

## 👍 Contribution
1. Fork it
2. Create your feature branch (git checkout -b my-new-feature)
3. Commit your changes (git commit -m 'Add some feature')
4. Push to the branch (git push origin my-new-feature)
5. Create new Pull Request


CHANGELOG.md
## [5.1.0]

- Update example code with gradient feature
- Add gradient colors support (#73)

## [5.0.1]

- [MacOS] Add macos support in example
- <fix> #66 Pie chart not animating with version 5.0.0+ (#69)  
- Accept center text style as a paramter

## [5.0.0]

* Show custom color when pie chart is empty.
* Set Colors.grey as default value for emptyColor.
* Update README.md with decimalPlaces docs.
* [Null Safety] Migrate project to null safety.

## [4.0.1]

* Fix README.md for broken images link

## [4.0.0] **Breaking**

* Wrap different parameters in chartValuesOptions, legendOptions
* Fix number of issues

## [3.1.1]

* Fix [#22](https://github.com/apgapg/caviare_flutter_pie_chart/issues/22) showLegends = false causes exception.

## [3.1.0]

* Add ring shape pie chart support

## [3.0.0] **Breaking**

* Migrate to AndroidX
* Add dark mode theme
* Remove color from default legendStyle to support dark mode
* Add enum LegendPosition to align chart legend **Breaking**
* Restructure whole project. Optimize code. Code Cleanup

Thanks [@xsahil03x](https://github.com/xsahil03x) for these changes

## [2.0.0]

* Fix padding, margin **Breaking**
* Fix error when parent is row
* Adapt to screen size **Breaking**
* Override chartRadius when screen is smaller compare to chartRadius **Breaking**
* Update example with macos and web support

## [1.3.0]

* Center text to sector center-line
* Add text labels

## [1.2.0]

* Merge [#12](https://github.com/apgapg/caviare_flutter_pie_chart/issues/12) Added option for showing decimal places in chart values.
  Thanks [@VB10](https://github.com/VB10) for PR

## [1.1.0]

* Merge [#11](https://github.com/apgapg/caviare_flutter_pie_chart/issues/11) Added option for changing initial angle of pie chart.
  Thanks [@mschneider](https://github.com/mschneider) for PR

## [1.0.0]

* Added [#8](https://github.com/apgapg/caviare_flutter_pie_chart/issues/8) Added option for showing chart values outside the pie
  chart [@guyzk](https://github.com/guyzk)
* Added [#7](https://github.com/apgapg/caviare_flutter_pie_chart/issues/7) Hide 0 values on pie chart.
  Thanks [@guyzk](https://github.com/https://github.com/guyzk)
* Added legend fontFamily support. Thanks [@MahdiPishguy](https://github.com/MahdiPishguy) for PR

## [0.9.0]

* Fixes [#5](https://github.com/apgapg/caviare_flutter_pie_chart/issues/5) Added custom colorlist optional parameter.
  Thanks [@SJente](https://github.com/SJente)
* Fixes [#3](https://github.com/apgapg/caviare_flutter_pie_chart/issues/3) Update PieChart when data changes.
  Thanks [@leehuwuj](https://github.com/https://github.com/leehuwuj)
* Fixes [#2](https://github.com/apgapg/caviare_flutter_pie_chart/issues/2) Added showLegends bool as optional Parameter.
  Thanks [@cuitao1988](https://github.com/https://github.com/cuitao1988)
* Fixes Exception when data length was greater than 5
* Added key to PieChart

## [0.8.0]

* Update README.md
* Format code
* Update min Dart SDK version  
  Thanks [@xsahil03x](https://github.com/xsahil03x) for this PR

## [0.7.0] - Initial Release


test/pie_chart_test.dart
import 'package:caviare_flutter_pie_chart/pie_chart.dart';
import 'package:caviare_flutter_pie_chart/src/chart_painter.dart';
import 'package:caviare_flutter_pie_chart/src/legend.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  final dataMap = <String, double>{};
  dataMap.putIfAbsent("Flutter", () => 5);
  dataMap.putIfAbsent("React", () => 3);
  dataMap.putIfAbsent("Xamarin", () => 2);
  dataMap.putIfAbsent("Ionic", () => 2);

  testWidgets('Chart is present', (WidgetTester tester) async {
    await tester.pumpWidget(MaterialApp(
      home: Scaffold(
        body: Center(
          child: PieChart(
            dataMap: dataMap,
          ),
        ),
      ),
    ));
    expect(
      find.byWidgetPredicate(
        (Widget widget) =>
            widget is CustomPaint && widget.painter is PieChartPainter,
      ),
      findsOneWidget,
    );
  });
  group("Legend(s) visibility test", () {
    testWidgets('Legend(s) is visibile', (WidgetTester tester) async {
      await tester.pumpWidget(MaterialApp(
        home: Scaffold(
          body: Center(
            child: PieChart(
              dataMap: dataMap,
            ),
          ),
        ),
      ));
      expect(
        find.byType(Legend),
        findsWidgets,
      );
    });
    testWidgets('Legend(s) number is correct', (WidgetTester tester) async {
      await tester.pumpWidget(MaterialApp(
        home: Scaffold(
          body: Center(
            child: PieChart(
              dataMap: dataMap,
            ),
          ),
        ),
      ));
      expect(
        find.byType(Legend),
        findsNWidgets(dataMap.length),
      );
    });
    testWidgets('Legend(s) is hidden', (WidgetTester tester) async {
      await tester.pumpWidget(MaterialApp(
        home: Scaffold(
          body: Center(
            child: PieChart(
              dataMap: dataMap,
              legendOptions: LegendOptions(showLegends: false),
            ),
          ),
        ),
      ));
      expect(
        find.byType(Legend),
        findsNothing,
      );
    });
  });

  testWidgets('Chart Data is correct', (WidgetTester tester) async {
    final falseDataMap = <String, dynamic>{};
    falseDataMap.putIfAbsent("Flutter", () => 5);
    falseDataMap.putIfAbsent("React", () => 3);
    falseDataMap.putIfAbsent("Xamarin", () => 2);
    falseDataMap.putIfAbsent("Ionic", () => 2);

    expectLater(
      () async {
        await tester.pumpWidget(MaterialApp(
          home: Scaffold(
            body: Center(
              child: PieChart(
                dataMap: falseDataMap as Map<String, double>,
              ),
            ),
          ),
        ));
      },
      throws,
    );
  });

  test("Test color return when index is greater than color list", () {
    final List<Color> colorList = [
      Color(0xFFff7675),
      Color(0xFF74b9ff),
      Color(0xFF55efc4),
      Color(0xFFffeaa7),
    ];
    expect(
      getColor(colorList, 5).runtimeType,
      Color,
    );
  });
}


lib/src/chart_painter.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:caviare_flutter_pie_chart/pie_chart.dart';

class PieChartPainter extends CustomPainter {
  List<Paint> _paintList = [];
  late List<double> _subParts;
  List<String>? _subTitles;
  double _total = 0;
  double _totalAngle = math.pi * 2;

  final TextStyle? chartValueStyle;
  final Color? chartValueBackgroundColor;
  final double? initialAngle;
  final bool? showValuesInPercentage;
  final bool showChartValues;
  final bool showChartValuesOutside;
  final int? decimalPlaces;
  final bool? showChartValueLabel;
  final ChartType? chartType;
  final String? centerText;
  final TextStyle? centerTextStyle;
  final Function? formatChartValues;
  final double? strokeWidth;
  final Color? emptyColor;
  final List<List<Color>>? gradientList;
  final List<Color>? emptyColorGradient;

  double _prevAngle = 0;

  PieChartPainter(double angleFactor, this.showChartValues,
      this.showChartValuesOutside, List<Color> colorList,
      {this.chartValueStyle,
      this.chartValueBackgroundColor,
      required List<double> values,
      List<String>? titles,
      this.initialAngle,
      this.showValuesInPercentage,
      this.decimalPlaces,
      this.showChartValueLabel,
      this.chartType,
      this.centerText,
      this.centerTextStyle,
      this.formatChartValues,
      this.strokeWidth,
      this.emptyColor,
      this.gradientList,
      this.emptyColorGradient}) {
    _total = values.fold(0, (v1, v2) => v1 + v2);
    if (gradientList?.isEmpty ?? true) {
      for (int i = 0; i < values.length; i++) {
        final paint = Paint()..color = getColor(colorList, i);
        if (chartType == ChartType.ring) {
          paint.style = PaintingStyle.stroke;
          paint.strokeWidth = strokeWidth!;
        }
        _paintList.add(paint);
      }
    }
    _totalAngle = angleFactor * math.pi * 2;
    _subParts = values;
    _subTitles = titles;
  }

  @override
  void paint(Canvas canvas, Size size) {
    var side = size.width < size.height ? size.width : size.height;
    side -= 50;
    if (_total == 0) {
      final paint = Paint()..color = emptyColor!;
      if (chartType == ChartType.ring) {
        paint.style = PaintingStyle.stroke;
        paint.strokeWidth = strokeWidth!;
      }
      canvas.drawArc(
        new Rect.fromLTWH(25, 25, side, side),
        _prevAngle,
        360,
        chartType == ChartType.disc ? true : false,
        paint,
      );
    } else {
      final paint = Paint()..color = emptyColor!;
      if (chartType == ChartType.ring) {
        paint.style = PaintingStyle.stroke;
        paint.strokeWidth = strokeWidth!;
      }
      canvas.drawArc(
        new Rect.fromLTWH(25, 25, side, side),
        _prevAngle,
        360,
        chartType == ChartType.disc ? true : false,
        paint,
      );
      final paint2 = Paint()..color = Color(0xFF594F5E);
      if (chartType == ChartType.ring) {
        paint2.style = PaintingStyle.stroke;
        paint2.strokeWidth = 2;
      }
      canvas.drawArc(
        new Rect.fromLTWH(25 - 10, 25 - 10, side + 20, side + 20),
        _prevAngle,
        360,
        chartType == ChartType.disc ? true : false,
        paint2,
      );
      for (int i = 0; i < 8; i++) {
        final radius = showChartValuesOutside ? (side / 2) - 20 : side / 3;
        final x = (radius) * math.cos(45 * i * math.pi / 180);
        final y = (radius) * math.sin(45 * i * math.pi / 180);
        final x2 = x + 25;
        final y2 = y + 25;
        _drawName(canvas, ((i * 3 + 6) % 24).toString(), x2, y2, side);
      }

      // final radius = showChartValuesOutside ? (side / 2) + 16 : side / 3;
      // final x = (radius) * math.cos(270 * math.pi / 180);
      // final y = (radius) * math.sin(270 * math.pi / 180);
      // final x2 = x + 0;
      // final y2 = y + 0;
      // _drawName(canvas, (7 * 3 + 3).toString(), x2, y2,
      //     x >= 0 || y <= 0 ? side + 60 : side - 60);
      // _drawName(
      //     canvas, '24', x2, y2, x2 <= 0 || y2 <= 0 ? side + 80 : side - 80);

      _prevAngle = this.initialAngle! * math.pi / 180;
      final isGradientPresent = gradientList?.isNotEmpty ?? false;
      final isNonGradientElementPresent =
          (_subParts.length - (gradientList?.length ?? 0)) > 0;
      for (int i = 0; i < _subParts.length; i++) {
        if (isGradientPresent) {
          final Rect _boundingSquare = Rect.fromLTWH(25, 25, side, side);
          final _endAngle = (((_totalAngle) / _total) * _subParts[i]);
          final paint = Paint();

          final _normalizedPrevAngle = (_prevAngle - 0.15) % (math.pi * 2);
          final _normalizedEndAngle = (_endAngle + 0.15) % (math.pi * 2);
          final Gradient _gradient = SweepGradient(
            transform: GradientRotation(_normalizedPrevAngle),
            endAngle: _normalizedEndAngle,
            colors: getGradient(gradientList!, i,
                isNonGradientElementPresent: isNonGradientElementPresent,
                emptyColorGradient: emptyColorGradient!),
          );
          paint.shader = _gradient.createShader(_boundingSquare);
          if (chartType == ChartType.ring) {
            paint.style = PaintingStyle.stroke;
            paint.strokeWidth = strokeWidth!;
            paint.strokeCap = StrokeCap.round;
          }
          canvas.drawArc(
            _boundingSquare,
            _prevAngle,
            _endAngle,
            chartType == ChartType.disc ? true : false,
            paint,
          );
        } else {
          canvas.drawArc(
            new Rect.fromLTWH(25, 25, side, side),
            _prevAngle,
            (((_totalAngle) / _total) * _subParts[i]),
            chartType == ChartType.disc ? true : false,
            _paintList[i],
          );
        }
        final radius = showChartValuesOutside ? (side / 2) + 16 : side / 3;
        final cosx = ((((_totalAngle) / _total) * _subParts[i]) / 2);
        final cosy = ((((_totalAngle) / _total) * _subParts[i]) / 2);
        final x = (radius) * math.cos(_prevAngle + cosx);
        final y = (radius) * math.sin(_prevAngle + cosy);
        if (_subParts.elementAt(i) > 0) {
          final value = formatChartValues != null
              ? formatChartValues!(_subParts.elementAt(i))
              : _subParts.elementAt(i).toStringAsFixed(this.decimalPlaces!);

          if (showChartValues) {
            final name = showValuesInPercentage!
                ? (((_subParts.elementAt(i) / _total) * 100)
                        .toStringAsFixed(this.decimalPlaces!) +
                    '%')
                : value;
            final x2 = x + 20;
            final y2 = y + 20;
            // _drawName(
            //     canvas, name, x2, y2, x <= 0 || y <= 0 ? side + 60 : side - 60);
          }
        }
        _prevAngle = _prevAngle + (((_totalAngle) / _total) * _subParts[i]);
      }
    }

    if (centerText != null && centerText!.trim().isNotEmpty) {
      _drawCenterText(canvas, side);
    }
  }

  void _drawCenterText(Canvas canvas, double side) {
    _drawName(canvas, centerText, 0, 0, side, style: centerTextStyle);
  }

  void _drawName(
    Canvas canvas,
    String? name,
    double x,
    double y,
    double side, {
    TextStyle? style,
  }) {
    TextSpan span = TextSpan(
      style: style ?? chartValueStyle,
      text: name,
    );
    TextPainter tp = TextPainter(
      text: span,
      textAlign: TextAlign.center,
      textDirection: TextDirection.ltr,
    );
    tp.layout();

    if (showChartValueLabel!) {
      //Draw text background box
      final rect = Rect.fromCenter(
        center: Offset((side / 2 + x), (side / 2 + y)),
        width: tp.width + 6,
        height: tp.height + 4,
      );
      final rRect = RRect.fromRectAndRadius(rect, Radius.circular(4));
      final paint = Paint()
        ..color = chartValueBackgroundColor ?? Colors.grey[200]!
        ..style = PaintingStyle.fill;
      canvas.drawRRect(rRect, paint);
    }
    //Finally paint the text above box
    tp.paint(
      canvas,
      new Offset(
        (side / 2 + x) - (tp.width / 2),
        (side / 2 + y) - (tp.height / 2),
      ),
    );
  }

  @override
  bool shouldRepaint(PieChartPainter oldDelegate) =>
      oldDelegate._totalAngle != _totalAngle;
}


example/lib/main.dart
import 'package:caviare_flutter_pie_chart/pie_chart.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Pie Chart Demo',
      theme: ThemeData(
        primarySwatch: Colors.blueGrey,
      ),
      darkTheme: ThemeData(
        primarySwatch: Colors.blueGrey,
        brightness: Brightness.dark,
      ),
      home: HomePage(),
    );
  }
}

enum LegendShape { Circle, Rectangle }

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  final dataMap = <String, double>{
    "Flutter": 5,
    "React": 3,
    "Xamarin": 2,
    "Ionic": 2,
  };
  final colorList = <Color>[
    Color(0xfffdcb6e),
    Color(0xff0984e3),
    Color(0xfffd79a8),
    Color(0xffe17055),
    Color(0xff6c5ce7),
  ];

  final gradientList = <List<Color>>[
    [
      Color.fromRGBO(223, 250, 92, 1),
      Color.fromRGBO(129, 250, 112, 1),
    ],
    [
      Color.fromRGBO(129, 182, 205, 1),
      Color.fromRGBO(91, 253, 199, 1),
    ],
    [
      Color.fromRGBO(175, 63, 62, 1.0),
      Color.fromRGBO(254, 154, 92, 1),
    ]
  ];

  ChartType? _chartType = ChartType.disc;
  bool _showCenterText = true;
  double? _ringStrokeWidth = 32;
  double? _chartLegendSpacing = 32;

  bool _showLegendsInRow = false;
  bool _showLegends = false;

  bool _showChartValueBackground = true;
  bool _showChartValues = true;
  bool _showChartValuesInPercentage = false;
  bool _showChartValuesOutside = false;

  bool _showGradientColors = false;

  LegendShape? _legendShape = LegendShape.Circle;
  LegendPosition? _legendPosition = LegendPosition.right;

  int key = 0;

  @override
  Widget build(BuildContext context) {
    final chart = PieChart(
      key: ValueKey(key),
      dataMap: dataMap,
      animationDuration: Duration(milliseconds: 800),
      chartLegendSpacing: _chartLegendSpacing!,
      chartRadius: MediaQuery.of(context).size.width / 3.2 > 300
          ? 300
          : MediaQuery.of(context).size.width / 3.2,
      colorList: colorList,
      initialAngleInDegree: 0,
      chartType: _chartType!,
      centerText: _showCenterText ? "HYBRID" : null,
      legendOptions: LegendOptions(
        showLegendsInRow: _showLegendsInRow,
        legendPosition: _legendPosition!,
        showLegends: _showLegends,
        legendShape: _legendShape == LegendShape.Circle
            ? BoxShape.circle
            : BoxShape.rectangle,
        legendTextStyle: TextStyle(
          fontWeight: FontWeight.bold,
        ),
      ),
      chartValuesOptions: ChartValuesOptions(
        showChartValueBackground: _showChartValueBackground,
        showChartValues: _showChartValues,
        showChartValuesInPercentage: _showChartValuesInPercentage,
        showChartValuesOutside: _showChartValuesOutside,
      ),
      ringStrokeWidth: _ringStrokeWidth!,
      emptyColor: Colors.grey,
      gradientList: _showGradientColors ? gradientList : null,
      emptyColorGradient: [
        Color(0xff6c5ce7),
        Colors.blue,
      ],
    );
    final settings = SingleChildScrollView(
      child: Card(
        margin: EdgeInsets.all(12),
        child: Column(
          children: [
            SwitchListTile(
              value: _showGradientColors,
              title: Text("Show Gradient Colors"),
              onChanged: (val) {
                setState(() {
                  _showGradientColors = val;
                });
              },
            ),
            ListTile(
              title: Text(
                'Pie Chart Options'.toUpperCase(),
                style: Theme.of(context).textTheme.overline!.copyWith(
                      fontSize: 12,
                      fontWeight: FontWeight.bold,
                    ),
              ),
            ),
            ListTile(
              title: Text("chartType"),
              trailing: Padding(
                padding: const EdgeInsets.symmetric(horizontal: 16.0),
                child: DropdownButton<ChartType>(
                  value: _chartType,
                  items: [
                    DropdownMenuItem(
                      child: Text("disc"),
                      value: ChartType.disc,
                    ),
                    DropdownMenuItem(
                      child: Text("ring"),
                      value: ChartType.ring,
                    ),
                  ],
                  onChanged: (val) {
                    setState(() {
                      _chartType = val;
                    });
                  },
                ),
              ),
            ),
            ListTile(
              title: Text("ringStrokeWidth"),
              trailing: Padding(
                padding: const EdgeInsets.symmetric(horizontal: 16.0),
                child: DropdownButton<double>(
                  value: _ringStrokeWidth,
                  disabledHint: Text("select chartType.ring"),
                  items: [
                    DropdownMenuItem(
                      child: Text("16"),
                      value: 16,
                    ),
                    DropdownMenuItem(
                      child: Text("32"),
                      value: 32,
                    ),
                    DropdownMenuItem(
                      child: Text("48"),
                      value: 48,
                    ),
                  ],
                  onChanged: (_chartType == ChartType.ring)
                      ? (val) {
                          setState(() {
                            _ringStrokeWidth = val;
                          });
                        }
                      : null,
                ),
              ),
            ),
            SwitchListTile(
              value: _showCenterText,
              title: Text("showCenterText"),
              onChanged: (val) {
                setState(() {
                  _showCenterText = val;
                });
              },
            ),
            ListTile(
              title: Text("chartLegendSpacing"),
              trailing: Padding(
                padding: const EdgeInsets.symmetric(horizontal: 16.0),
                child: DropdownButton<double>(
                  value: _chartLegendSpacing,
                  disabledHint: Text("select chartType.ring"),
                  items: [
                    DropdownMenuItem(
                      child: Text("16"),
                      value: 16,
                    ),
                    DropdownMenuItem(
                      child: Text("32"),
                      value: 32,
                    ),
                    DropdownMenuItem(
                      child: Text("48"),
                      value: 48,
                    ),
                    DropdownMenuItem(
                      child: Text("64"),
                      value: 64,
                    ),
                  ],
                  onChanged: (val) {
                    setState(() {
                      _chartLegendSpacing = val;
                    });
                  },
                ),
              ),
            ),
            ListTile(
              title: Text(
                'Legend Options'.toUpperCase(),
                style: Theme.of(context).textTheme.overline!.copyWith(
                      fontSize: 12,
                      fontWeight: FontWeight.bold,
                    ),
              ),
            ),
            SwitchListTile(
              value: _showLegendsInRow,
              title: Text("showLegendsInRow"),
              onChanged: (val) {
                setState(() {
                  _showLegendsInRow = val;
                });
              },
            ),
            SwitchListTile(
              value: _showLegends,
              title: Text("showLegends"),
              onChanged: (val) {
                setState(() {
                  _showLegends = val;
                });
              },
            ),
            ListTile(
              title: Text("legendShape"),
              trailing: Padding(
                padding: const EdgeInsets.symmetric(horizontal: 16.0),
                child: DropdownButton<LegendShape>(
                  value: _legendShape,
                  items: [
                    DropdownMenuItem(
                      child: Text("BoxShape.circle"),
                      value: LegendShape.Circle,
                    ),
                    DropdownMenuItem(
                      child: Text("BoxShape.rectangle"),
                      value: LegendShape.Rectangle,
                    ),
                  ],
                  onChanged: (val) {
                    setState(() {
                      _legendShape = val;
                    });
                  },
                ),
              ),
            ),
            ListTile(
              title: Text("legendPosition"),
              trailing: Padding(
                padding: const EdgeInsets.symmetric(horizontal: 16.0),
                child: DropdownButton<LegendPosition>(
                  value: _legendPosition,
                  items: [
                    DropdownMenuItem(
                      child: Text("left"),
                      value: LegendPosition.left,
                    ),
                    DropdownMenuItem(
                      child: Text("right"),
                      value: LegendPosition.right,
                    ),
                    DropdownMenuItem(
                      child: Text("top"),
                      value: LegendPosition.top,
                    ),
                    DropdownMenuItem(
                      child: Text("bottom"),
                      value: LegendPosition.bottom,
                    ),
                  ],
                  onChanged: (val) {
                    setState(() {
                      _legendPosition = val;
                    });
                  },
                ),
              ),
            ),
            ListTile(
              title: Text(
                'Chart values Options'.toUpperCase(),
                style: Theme.of(context).textTheme.overline!.copyWith(
                      fontSize: 12,
                      fontWeight: FontWeight.bold,
                    ),
              ),
            ),
            SwitchListTile(
              value: _showChartValueBackground,
              title: Text("showChartValueBackground"),
              onChanged: (val) {
                setState(() {
                  _showChartValueBackground = val;
                });
              },
            ),
            SwitchListTile(
              value: _showChartValues,
              title: Text("showChartValues"),
              onChanged: (val) {
                setState(() {
                  _showChartValues = val;
                });
              },
            ),
            SwitchListTile(
              value: _showChartValuesInPercentage,
              title: Text("showChartValuesInPercentage"),
              onChanged: (val) {
                setState(() {
                  _showChartValuesInPercentage = val;
                });
              },
            ),
            SwitchListTile(
              value: _showChartValuesOutside,
              title: Text("showChartValuesOutside"),
              onChanged: (val) {
                setState(() {
                  _showChartValuesOutside = val;
                });
              },
            ),
          ],
        ),
      ),
    );
    return Scaffold(
      appBar: AppBar(
        title: Text("Pie Chart @apgapg"),
        actions: [
          ElevatedButton(
            onPressed: () {
              setState(() {
                key = key + 1;
              });
            },
            child: Text("Reload".toUpperCase()),
          ),
        ],
      ),
      body: LayoutBuilder(
        builder: (_, constraints) {
          if (constraints.maxWidth >= 600) {
            return Row(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Flexible(
                  flex: 3,
                  fit: FlexFit.tight,
                  child: chart,
                ),
                Flexible(
                  flex: 2,
                  fit: FlexFit.tight,
                  child: settings,
                )
              ],
            );
          } else {
            return SingleChildScrollView(
              child: Column(
                children: [
                  Container(
                    child: chart,
                    margin: EdgeInsets.symmetric(
                      vertical: 32,
                    ),
                  ),
                  settings,
                ],
              ),
            );
          }
        },
      ),
    );
  }
}
